src/bigint.rs

```diff
@@ -467,146 +522,303 @@ impl Unsigned for BigUint {}
+    fn add(mut self, other: &[BigDigit]) -> BigUint {
+        let new_len = cmp::max(self.data.len(), other.len()) + 1;
 
-        if other.data.len() > sum.len() {
-            let additional = other.data.len() - sum.len();
-            sum.reserve(additional);
-            sum.extend(repeat(ZERO_BIG_DIGIT).take(additional));
-        }
-        let other_iter = other.data.iter().cloned().chain(repeat(ZERO_BIG_DIGIT));
+        // resize isn't stable yet:
+        //self.data.resize(new_len, 0);
 
-        let mut carry = 0;
-        for (a, b) in sum.iter_mut().zip(other_iter) {
-            let d = (*a as DoubleBigDigit)
-                + (b as DoubleBigDigit)
-                + (carry as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            carry = hi;
-            *a = lo;
+        while self.data.len() < new_len {
+            self.data.push(0);
         }
 
-        if carry != 0 { sum.push(carry); }
-        BigUint::new(sum)
+        add2(&mut self.data[..], other);
+        self.normalize()
+    }
+}
+
+impl<'a> Add<&'a BigUint> for BigUint {
+    type Output = BigUint;
+
+    fn add(self, other: &BigUint) -> BigUint {
+        self + &other.data[..]
     }
 }
 
 forward_all_binop_to_val_ref!(impl Sub for BigUint, sub);
 
-impl<'a> Sub<&'a BigUint> for BigUint {
-    type Output = BigUint;
+fn sub2(a: &mut [BigDigit], b: &[BigDigit]) {
+    let mut b_iter = b.iter();
+    let mut borrow = 0;
 
-    fn sub(self, other: &BigUint) -> BigUint {
-        let mut diff = self.data;
-        let other = &other.data;
-        assert!(diff.len() >= other.len(), "arithmetic operation overflowed");
-
-        let mut borrow: DoubleBigDigit = 0;
-        for (a, &b) in diff.iter_mut().zip(other.iter()) {
-            let d = big_digit::BASE - borrow
-                + (*a as DoubleBigDigit)
-                - (b as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            /*
-            hi * (base) + lo == 1*(base) + ai - bi - borrow
-            => ai - bi - borrow < 0 <=> hi == 0
-             */
-            borrow = if hi == 0 { 1 } else { 0 };
-            *a = lo;
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = sbb(*ai, *bi, &mut borrow);
+        } else if borrow != 0 {
+            *ai = sbb(*ai, 0, &mut borrow);
+        } else {
+            break;
         }
+    }
 
-        for a in &mut diff[other.len()..] {
-            if borrow == 0 { break }
-            let d = big_digit::BASE - borrow
-                + (*a as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            borrow = if hi == 0 { 1 } else { 0 };
-            *a = lo;
-        }
+    /* note: we're _required_ to fail on underflow */
+    assert!(borrow == 0 && b_iter.all(|x| *x == 0),
+            "Cannot subtract b from a because b is larger than a.");
+}
+
+impl<'a> Sub<&'a [BigDigit]> for BigUint {
```

Comment:
Ditto on undesirable public API with slices -- but I see this one is used both in the regular `Sub` and in `sub_sign`.  Since the only thing it does over than `sub2` is normalize, let's just do that in the callers.  Or else make a new private function instead of this new `impl Sub`.